metalearn/metafeatures/base.py

```diff
@@ -0,0 +1,146 @@
+        problem_type: ProblemType,
+        groups: List[MetafeatureGroup],
+        argmap: Optional[Dict[str,str]] = {}
+    ) -> None:
+        """
+        Decorates ``computer``, a metafeature computing function
+        with metadata about that function.
+        
+        Parameters
+        ----------
+        computer
+            The function that computes the metafeatures.
+        returns
+            The names of the metafeatures that ``computer`` returns, specified in
+            the same order as ``computer`` returns them.
+        problem_type
+            The type of ML problem `computer`'s metafeatures can be computed for.
+        groups
+            The metafeature groups this computer's returned metafeatures belong to.
+            e.g. statistical, info-theoretic, simple, etc.
+        argmap
+            A custom map of ``computer``'s argument names to the global resource names
+            that will be passed as ``computer``'s arguments when ``computer`` is called.
+        """
+        super(MetafeatureComputer, self).__init__(computer, returns, argmap)
+        self.groups = groups
+        self.problem_type = problem_type
+
+
+class ResourceComputerMap(MutableMapping):
+    def __init__(self, computers: Union[ResourceComputer,List[ResourceComputer],None] = None) -> None:
+        """
+        Wraps a dictionary map of resource names to their computers.
+        Includes visibility into whether duplicate computers
+        are trying to become associated with a resource in the map e.g.
+        if a package developer has accidentally declared two computers
+        that return the same resource.
+        """
+        self._map: Dict[str,ResourceComputer] = {}
+        if computers is not None:
+            self.add(computers)
+    
+    def add(self, computers: Union[ResourceComputer,List[ResourceComputer]]) -> None:
+        """
+        Adds more resource name/resource computer key/value
+        pairs to a resource map, throwing an error on duplicates.
+        """
+        if isinstance(computers, list):
+            for computer in computers:
+                self._add_one(computer)
+        elif isinstance(computers, ResourceComputer):
+            self._add_one(computers)
+        else:
+            raise ValueError("computers must be ResourceComputer or List[ResourceComputer]")
+    
+    def __getitem__(self, key: str = None) -> ResourceComputer:
+        """Used for getting a resource from the map."""
+        return self._map[key]
+
+    def _add_one(self, computer: ResourceComputer) -> None:
+        if not isinstance(computer, ResourceComputer):
+            raise ValueError(f"computer is not a ResourceComputer; it is a {type(computer)}")
+            
+        for resource_name in computer.returns:
+            self.__setitem__(resource_name, computer)
+    
+    def __setitem__(self, resource_name: str, computer: ResourceComputer):
```

Comment:
This seems like a good thing to have. Since this is a developer facing API, we could keep this class for just this method and change the error to an assert.

Reply:
Sounds good, I just changed it to an assert in a recent commit, now pushed to remote here and viewable.